### PR TITLE
[AIRLFOW-5163] Default do_xcom_push to False

### PR DESCRIPTION
**Pull request by BasPH**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-5163
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

By default `BaseOperator.do_xcom_push` is set to `True`, pushing any return value of execute() to an XCom by default. This is most of the time not desirable and must be explicitly turned off. It makes more sense to set it to `False` by default.

In my memory there was a sentence about this behaviour in the docs but I cannot find it. If anybody knows, let me know.

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [ ] Passes `flake8`

### CODE REVIEW

**Comment by turbaszek** [comment]
@BasPH does this change requires any changes in example dags?

**Comment by BasPH** [comment]
`example_xcom.py` will fail because it has one task depending on an XCom from return value. I'll update it.

**Comment by kaxil** [comment]
I think this was done to keep the changes back-compatible. As pushing to xcom was the default behavior.

**Comment by turbaszek** [comment]
I think it will be more than one DAG, for instance many gcp examples uses returned values:
https://github.com/apache/airflow/blob/master/airflow/contrib/example_dags/example_gcp_natural_language.py

**Comment by kaxil** [comment]
We can mark this for 2.0. Can we add a note in XCom section in the docs too ?

**Comment by bolkedebruin** [comment]
Why does this need to be disabled? What is non desireable about it? The argumentation is flawed in the pr.

**Comment by BasPH** [comment]
I've seen multiple incidents where a return statement unintentionally pushed multiple GBs into the metastore. Communicating via XComs is always an intentional decision, therefore IMO the default behaviour would make much more sense if the execute() return value by default does nothing and only pushes to XCom if `do_xcom_push` is (intentionally) set to `True`. At the moment it feels counterintuitive.

@bolkedebruin does this clarify?

**Review by Fokko — APPROVED** [review]
I think it should be `False` by default. If this is True, it will push a lot of information to the metastore. Looking at the DAGs that I've written, I use xcom only in a small percentage of the time, so I think it should be a conscious choice.

File: UPDATING.md [hunk 1]
```diff
@@ -23,6 +23,12 @@ assists users migrating to a new version.
 
 ## Airflow Master
 
+### Operator return values are not pushed to XCom by default
```
Comment (kaxil) [inline UPDATING.md:44]: Can we make a note of this here as well -> https://github.com/apache/airflow/blob/master/docs/concepts.rst#xcoms

**Comment by mik-laj** [comment]
I was thinking about operators from the "Get" category whose main purpose is to download information to XCom eg. ``CloudMemorystoreGetInstanceOperator``. Should the user worry about an additional parameter when using them?

**Comment by potiuk** [comment]
This is quite drastic change.

While I agree it should be done, I think this change should also set do_xcom_push to `True` for all the operators that currently return a value from execute and do not have "do_xcom_push" set  (after checking that it's intentional). It should be rather safe and easy change.

I think it's not enough to write it in updating.md in this case. There will be weird, unrelated problems if they do_xcom_push is set to False for DAGs that rely on it. Often you will see missing or empty value in a task that is further upstream and I imagine this to be nightmare to debug for huge DAGs.

I did a quick search for operators that do not have *do_xcom_push* but they have some "return\s\S.*$" regexp. They might of course have false positives of course (but by quick look - not many). I think the list is large enough to worry, but small enough that such review and global change is actually doable in short time.

```
./airflow/sensors/base_sensor_operator.py
./airflow/operators/hive_stats_operator.py
./airflow/operators/druid_check_operator.py
./airflow/operators/presto_check_operator.py
./airflow/operators/pig_operator.py
./airflow/operators/http_operator.py
./airflow/operators/subdag_operator.py
./airflow/operators/hive_operator.py
./airflow/operators/python_operator.py
./airflow/operators/check_operator.py
./airflow/operators/docker_operator.py
./airflow/operators/branch_operator.py
./airflow/operators/s3_to_hive_operator.py
./airflow/operators/dagrun_operator.py
./airflow/operators/bash_operator.py
./airflow/contrib/operators/sagemaker_model_operator.py
./airflow/contrib/operators/adls_list_operator.py
./airflow/contrib/operators/azure_container_instances_operator.py
./airflow/contrib/operators/aws_athena_operator.py
./airflow/contrib/operators/sagemaker_tuning_operator.py
./airflow/contrib/operators/opsgenie_alert_operator.py
./airflow/contrib/operators/sagemaker_training_operator.py
./airflow/contrib/operators/qubole_check_operator.py
./airflow/contrib/operators/jira_operator.py
./airflow/contrib/operators/kubernetes_pod_operator.py
./airflow/contrib/operators/s3_to_sftp_operator.py
./airflow/contrib/operators/jenkins_job_trigger_operator.py
./airflow/contrib/operators/databricks_operator.py
./airflow/contrib/operators/grpc_operator.py
./airflow/contrib/operators/snowflake_operator.py
./airflow/contrib/operators/sftp_to_s3_operator.py
./airflow/contrib/operators/s3_list_operator.py
./airflow/contrib/operators/sftp_operator.py
./airflow/contrib/operators/aws_sqs_publish_operator.py
./airflow/contrib/operators/winrm_operator.py
./airflow/contrib/operators/sagemaker_endpoint_config_operator.py
./airflow/contrib/operators/sns_publish_operator.py
./airflow/contrib/operators/wasb_delete_blob_operator.py
./airflow/contrib/operators/s3_to_gcs_operator.py
./airflow/contrib/operators/ssh_operator.py
./airflow/contrib/operators/sagemaker_endpoint_operator.py
./airflow/contrib/operators/qubole_operator.py
./airflow/contrib/operators/sagemaker_transform_operator.py
./airflow/contrib/operators/emr_create_job_flow_operator.py
./airflow/contrib/operators/emr_add_steps_operator.py
./airflow/contrib/operators/ecs_operator.py
./airflow/contrib/operators/awsbatch_operator.py
```

**Review by potiuk — CHANGES_REQUESTED** [review]
See the comment - I think we should fix all the current operators that intentionally rely on do_xcom_push = True by default..

**Comment by potiuk** [comment]
I can help with it @BasPH BTW -> we can split in half for example :)

**Comment by Fokko** [comment]
@bolkedebruin I think it should be set to `False` by default, for two reasons:
- It will potentially pollute the database by inserting xcom data, while not using it. This should be a deliberate choice. Most of the DAGs won't use xcom.
- If you by accident push a big object, it will throw an error because the payload is bigger than 48k: https://github.com/apache/airflow/blob/master/airflow/models/xcom.py#L36

@potiuk Thanks for figuring out which operators return data to xcom. However, if something is being pushed to xcom, it doesn't mean that it is being used somewhere downstream or in another dag. I believe we should go through the list and check where it actually makes sense to push it to xcom. I'm happy to help on this as well :-)

**Pull request closed without merging**